// next-campaign-page-kit/lib/actions/init.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const https = require('https');
const { execSync } = require('child_process');
const logger = require('../logger');

const D = '\x1b[90m';
const R = '\x1b[0m';

const CAMPAIGNS_JSON = JSON.stringify({}, null, 2) + '\n';

const SCRIPTS = {
    'setup': 'campaign-init',
    'start': 'campaign-start',
    'dev': 'campaign-dev',
    'build': 'campaign-build',
    'clone': 'campaign-clone',
    'config': 'campaign-config',
    'compress': 'campaign-compress',
    'compress:preview': 'campaign-compress --preview',
    'migrate': 'campaign-migrate',
};

const REPO = 'NextCommerceCo/campaign-cart-starter-templates';
const REF = 'main';
const TEMPLATES_URL = `https://raw.githubusercontent.com/${REPO}/${REF}/templates.json`;
const CAMPAIGNS_URL = `https://raw.githubusercontent.com/${REPO}/${REF}/_data/campaigns.json`;
const TARBALL_URL  = `https://codeload.github.com/${REPO}/tar.gz/refs/heads/${REF}`;
const AI_CONTEXT_DOC_URL = `https://raw.githubusercontent.com/${REPO}/${REF}/docs/campaign-page-kit-template-context.md`;

// Exit codes — every error path maps to one of these. 0 is reserved for success.
const EXIT = {
    OK: 0,
    GENERIC: 1,
    UPSTREAM_TEMPLATE_NOT_FOUND: 2,
    CONFLICT: 3,
    UPSTREAM_FETCH_FAILED: 4,
    MISSING_INPUT: 5,
    INVALID_INPUT: 6,
    PARTIAL_WRITE_ROLLED_BACK: 7,
    PARTIAL_WRITE_DIRTY: 8,
};

// AI context destinations per tool. `null` = don't write anything.
// `frontmatter` is prepended verbatim above the sentinel for tools that need it.
const AI_CONTEXT_TARGETS = {
    claude:  { path: 'CLAUDE.md',                            frontmatter: null },
    codex:   { path: 'AGENTS.md',                            frontmatter: null },
    cursor:  { path: '.cursor/rules/campaign-page-kit.mdc',  frontmatter: '---\ndescription: Next Commerce Campaign Page Kit context\nalwaysApply: true\n---\n' },
    copilot: { path: '.github/copilot-instructions.md',      frontmatter: null },
    none:    null,
};

const AI_CONTEXT_SENTINEL = [
    '<!-- generated by next-campaign-page-kit (campaign-init); upstream: docs/campaign-page-kit-template-context.md -->',
    '<!-- this file is overwritten on re-run of campaign-init; pass --keep-ai-context to preserve hand edits -->',
].join('\n');

// CLI flag schema. Each entry: { type: 'string'|'boolean', enum?: [...], target?: 'camelKey' }.
// Targets default to camelCase of the kebab key.
const FLAG_SCHEMA = {
    'template':         { type: 'string' },
    'slug':             { type: 'string' },
    'name':             { type: 'string' },
    'api-key':          { type: 'string' },
    'non-interactive':  { type: 'boolean' },
    'json':             { type: 'boolean' },
    'dry-run':          { type: 'boolean' },
    'overwrite':        { type: 'boolean' },
    'ai-context':       { type: 'string', enum: ['claude', 'codex', 'cursor', 'copilot', 'none'] },
    'keep-ai-context':  { type: 'boolean' },
    'help':             { type: 'boolean' },
    'h':                { type: 'boolean', target: 'help' },
};

// ---------------------------------------------------------------------------
// HTTP helpers
// ---------------------------------------------------------------------------

function fetchBuffer(url, redirects = 5) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                res.resume();
                if (redirects <= 0) return reject(new Error('too many redirects'));
                return resolve(fetchBuffer(res.headers.location, redirects - 1));
            }
            if (res.statusCode !== 200) {
                res.resume();
                return reject(new Error(`HTTP ${res.statusCode} fetching ${url}`));
            }
            const chunks = [];
            res.on('data', (c) => chunks.push(c));
            res.on('end', () => resolve(Buffer.concat(chunks)));
            res.on('error', reject);
        }).on('error', reject);
    });
}

async function fetchJson(url) {
    const buf = await fetchBuffer(url);
    return JSON.parse(buf.toString('utf8'));
}

// ---------------------------------------------------------------------------
// Template extraction (tarball → directory)
// ---------------------------------------------------------------------------

// Extract a single template's src/<slug>/ subtree from the upstream tarball
// using the system `tar` binary. No homemade parser, no truncation surprises.
function extractTemplate(tarball, slug, destDir) {
    const parent = path.dirname(destDir);
    fs.mkdirSync(parent, { recursive: true });
    const tmpDir = fs.mkdtempSync(path.join(parent, '.cpk-extract-'));
    try {
        execSync('tar -xz', { input: tarball, cwd: tmpDir, stdio: ['pipe', 'ignore', 'pipe'] });
        const repoRoot = fs.readdirSync(tmpDir).find(d =>
            fs.statSync(path.join(tmpDir, d)).isDirectory()
        );
        if (!repoRoot) throw new Error('tarball had no top-level directory');
        const sourceDir = path.join(tmpDir, repoRoot, 'src', slug);
        if (!fs.existsSync(sourceDir)) {
            throw new Error(`upstream has no src/${slug}/`);
        }
        fs.cpSync(sourceDir, destDir, { recursive: true });
        return countFiles(destDir);
    } finally {
        fs.rmSync(tmpDir, { recursive: true, force: true });
    }
}

function countFiles(dir) {
    let n = 0;
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        if (entry.isDirectory()) n += countFiles(path.join(dir, entry.name));
        else n++;
    }
    return n;
}

function replaceDirectoryWithRollback(stagedDir, destDir, commit) {
    const parent = path.dirname(destDir);
    const backupPrefix = path.join(parent, `.cpk-backup-${path.basename(destDir)}-`);
    let backupDir = null;
    let stagedMoved = false;
    let committed = false;

    try {
        if (fs.existsSync(destDir)) {
            backupDir = fs.mkdtempSync(backupPrefix);
            fs.rmSync(backupDir, { recursive: true, force: true });
            fs.renameSync(destDir, backupDir);
        }

        fs.renameSync(stagedDir, destDir);
        stagedMoved = true;
        commit();
        committed = true;

        if (backupDir) {
            try {
                fs.rmSync(backupDir, { recursive: true, force: true });
            } catch {
                // The install is already committed; avoid rolling back a valid install
                // because cleanup of the rollback backup failed.
            }
        }
    } catch (err) {
        if (committed) {
            throw err;
        }
        if (stagedMoved && fs.existsSync(destDir)) {
            fs.rmSync(destDir, { recursive: true, force: true });
        }
        if (backupDir && fs.existsSync(backupDir) && !fs.existsSync(destDir)) {
            fs.renameSync(backupDir, destDir);
        }
        throw err;
    } finally {
        if (!stagedMoved && fs.existsSync(stagedDir)) {
            fs.rmSync(stagedDir, { recursive: true, force: true });
        }
    }
}

// ---------------------------------------------------------------------------
// Pure helpers (used in main + exported for tests)
// ---------------------------------------------------------------------------

// Convert a free-form name into a valid campaign slug.
function slugify(s) {
    return String(s || '').toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-+|-+$/g, '');
}

// Filter + sort a templates.json manifest into the order the picker shows.
// Drops `hidden`, sorts by `priority` desc, name asc as tiebreak.
function selectableTemplates(manifest) {
    return (manifest && manifest.templates || [])
        .filter(t => !t.hidden)
        .sort((a, b) => {
            const dp = (b.priority || 0) - (a.priority || 0);
            return dp !== 0 ? dp : String(a.name || a.slug).localeCompare(String(b.name || b.slug));
        });
}

// Produce the next state of `_data/campaigns.json` after merging an upstream
// entry under `localSlug` and overriding the name with what the user provided.
function mergeCampaignEntry(localCampaigns, localSlug, upstreamEntry, name) {
    const base = upstreamEntry || { description: '' };
    return { ...localCampaigns, [localSlug]: { ...base, name } };
}

// Merge missing CLI scripts into a package.json object. Returns
// { pkg, added: [keys] }; does not mutate the input.
function mergeScripts(pkg) {
    const next = { ...pkg, scripts: { ...(pkg && pkg.scripts || {}) } };
    const added = [];
    for (const [k, v] of Object.entries(SCRIPTS)) {
        if (!next.scripts[k]) {
            next.scripts[k] = v;
            added.push(k);
        }
    }
    return { pkg: next, added };
}

// Apply an API key to a config.js source string. Returns the new string,
// or null if no `apiKey:` field was found.
function applyApiKey(configSource, apiKey) {
    const updated = configSource.replace(/apiKey:\s*['"].*?['"]/, `apiKey: '${apiKey.trim()}'`);
    return updated === configSource ? null : updated;
}

// kebab-case → camelCase: 'ai-context' → 'aiContext'.
function kebabToCamel(s) {
    return String(s).replace(/-([a-z0-9])/g, (_, c) => c.toUpperCase());
}

// Validate a campaign slug. Returns an error message string or null.
function validateSlug(s) {
    const t = String(s || '').trim();
    if (!t) return 'slug cannot be empty';
    if (!/^[a-z0-9][a-z0-9-]*$/.test(t)) return 'lowercase letters, digits, hyphens only (must start with a letter or digit)';
    return null;
}

// Look up the destination spec for an --ai-context tool name. Returns null
// for unknown tools and for `none` (the explicit "don't write" sentinel).
function aiContextTarget(tool) {
    if (!Object.prototype.hasOwnProperty.call(AI_CONTEXT_TARGETS, tool)) return null;
    return AI_CONTEXT_TARGETS[tool];
}

// Compose the final file content for an AI context destination: optional
// frontmatter, the sentinel, then the upstream document body.
function buildAiContextContent(upstreamBody, target) {
    const sentinelBlock = `${AI_CONTEXT_SENTINEL}\n\n`;
    return target.frontmatter
        ? `${target.frontmatter}\n${sentinelBlock}${upstreamBody}`
        : `${sentinelBlock}${upstreamBody}`;
}

// Parse argv against a flag schema. Returns { values, errors }.
// Supports `--key value`, `--key=value`, and `--flag` (boolean true).
// Unknown flags and missing required values produce structured errors.
function parseArgs(argv, schema = FLAG_SCHEMA) {
    const values = {};
    const errors = [];
    let i = 0;
    while (i < argv.length) {
        const tok = argv[i];
        // Accept --long-flags and -short-flags (e.g. -h). A bare `-` or `--`,
        // or a `-N` digit token, is treated as a positional argument and
        // rejected — we don't take any positional input.
        const isLong  = typeof tok === 'string' && tok.startsWith('--') && tok.length > 2;
        const isShort = typeof tok === 'string' && !tok.startsWith('--') && tok.startsWith('-') && tok.length > 1 && /^-[a-zA-Z]/.test(tok);
        if (!isLong && !isShort) {
            errors.push({ code: 'INVALID_INPUT', message: `unexpected positional argument: ${tok}` });
            i++;
            continue;
        }
        const prefixLen = isLong ? 2 : 1;
        const eq = tok.indexOf('=');
        const rawKey = eq === -1 ? tok.slice(prefixLen) : tok.slice(prefixLen, eq);
        const inlineVal = eq === -1 ? null : tok.slice(eq + 1);
        const def = schema[rawKey];
        if (!def) {
            errors.push({ code: 'INVALID_INPUT', message: `unknown flag: --${rawKey}` });
            // Skip a following value-like token so we don't double-report it.
            if (inlineVal === null && argv[i + 1] !== undefined && !argv[i + 1].startsWith('--')) i++;
            i++;
            continue;
        }
        const targetKey = def.target || kebabToCamel(rawKey);
        if (def.type === 'boolean') {
            if (inlineVal === null || inlineVal === '' || inlineVal === 'true') {
                values[targetKey] = true;
            } else if (inlineVal === 'false') {
                values[targetKey] = false;
            } else {
                errors.push({ code: 'INVALID_INPUT', message: `--${rawKey} is a boolean flag and does not accept a value` });
            }
            i++;
        } else { // string
            let v = inlineVal;
            if (v === null) {
                v = argv[i + 1];
                if (v === undefined || (typeof v === 'string' && v.startsWith('--'))) {
                    errors.push({ code: 'INVALID_INPUT', message: `--${rawKey} requires a value` });
                    i++;
                    continue;
                }
                i += 2;
            } else {
                i++;
            }
            if (def.enum && !def.enum.includes(v)) {
                errors.push({ code: 'INVALID_INPUT', message: `--${rawKey} must be one of: ${def.enum.join(', ')}` });
                continue;
            }
            values[targetKey] = v;
        }
    }
    return { values, errors };
}

// ---------------------------------------------------------------------------
// Result / output
// ---------------------------------------------------------------------------

function newResult({ dryRun, brandRoot }) {
    return {
        status: 'ok',
        dry_run: !!dryRun,
        brand_root: brandRoot,
        campaign: null,
        writes: {
            package_json_scripts_added: [],
            campaigns_json_seeded: false,
            campaigns_json_entry: null,
            files_extracted: 0,
            destination: null,
            api_key_applied: false,
            ai_context: [],
        },
        warnings: [],
        errors: [],
    };
}

function emitJson(result) {
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
}

// ---------------------------------------------------------------------------
// Help text
// ---------------------------------------------------------------------------

const HELP_TEXT = `campaign-init — scaffold a new campaign from a starter template

USAGE
  campaign-init                         interactive mode (prompts for everything)
  campaign-init [flags]                 pre-fills prompts from flags; prompts only for missing values
  campaign-init --non-interactive [flags]   never prompts; missing required input → exit 5
  campaign-init --json [flags]              emit a single JSON object on stdout (suppresses UI + prompts)

FLAGS
  --template <slug>          starter template slug (must exist in upstream templates.json)
  --slug <name>              local campaign slug (folder under src/, also URL path)
  --name <"Display Name">    campaign display name (defaults to upstream template name)
  --api-key <key>            Campaign Cart API key written to assets/config.js

  --non-interactive          never prompt; missing required input exits with code 5
  --json                     machine-readable stdout; suppresses all human UI
  --dry-run                  resolve plan without downloading the tarball or writing files
  --overwrite                replace existing src/<slug>/ + campaigns.json entry on conflict

  --ai-context <tool>        write upstream AI context doc for one of:
                               claude   → CLAUDE.md
                               codex    → AGENTS.md
                               cursor   → .cursor/rules/campaign-page-kit.mdc
                               copilot  → .github/copilot-instructions.md
                               none     → don't write (default)
  --keep-ai-context          do not overwrite an existing AI context file

  --help, -h                 show this help and exit 0

EXIT CODES
  0  success
  1  generic / unhandled error
  2  upstream template not found
  3  target conflict (src/<slug>/ or registry entry exists; use --overwrite)
  4  upstream fetch failed (templates.json, campaigns.json, tarball, or AI context doc)
  5  missing required input in --non-interactive mode
  6  invalid input (bad flag, bad slug, unknown tool, etc.)
  7  partial write rolled back (workspace clean)
  8  partial write failed to roll back (workspace dirty)

EXAMPLES
  # fully scripted scaffold for an agent:
  npx campaign-init --non-interactive --json \\
      --template olympus --slug grounding-mat-v2 \\
      --name "Grounding Mat V2" --api-key "$KEY" \\
      --ai-context claude
`;

// ---------------------------------------------------------------------------
// UI shim — clack in interactive mode, no-op in --json mode
// ---------------------------------------------------------------------------

function makeNoopUi() {
    const noop = () => {};
    return {
        intro: noop, outro: noop,
        log: { info: noop, success: noop, warn: noop, error: noop, step: noop },
        spinner: () => ({ start: noop, stop: noop, message: noop }),
        select: async () => null, confirm: async () => false, text: async () => null,
        isCancel: () => false,
    };
}

function makeClackUi(clack) {
    return {
        intro: clack.intro,
        outro: clack.outro,
        log: clack.log,
        spinner: clack.spinner,
        select: clack.select,
        confirm: clack.confirm,
        text: clack.text,
        isCancel: clack.isCancel,
    };
}

// ---------------------------------------------------------------------------
// AI context file write
// ---------------------------------------------------------------------------

async function writeAiContext({ brandRoot, tool, keep, dryRun }) {
    const target = aiContextTarget(tool);
    if (!target) return null;

    const fullPath = path.join(brandRoot, target.path);
    const exists = fs.existsSync(fullPath);

    if (exists && keep) {
        return { path: target.path, action: 'skipped', reason: '--keep-ai-context' };
    }
    if (dryRun) {
        return { path: target.path, action: exists ? 'would-overwrite' : 'would-create', reason: '--dry-run' };
    }

    let upstreamBody;
    try {
        const buf = await fetchBuffer(AI_CONTEXT_DOC_URL);
        upstreamBody = buf.toString('utf8');
    } catch (err) {
        const wrapped = new Error(`could not fetch AI context doc: ${err.message}`);
        wrapped.code = 'UPSTREAM_FETCH_FAILED';
        throw wrapped;
    }

    const content = buildAiContextContent(upstreamBody, target);
    fs.mkdirSync(path.dirname(fullPath), { recursive: true });
    fs.writeFileSync(fullPath, content, 'utf8');
    return { path: target.path, action: exists ? 'overwritten' : 'created' };
}

// ---------------------------------------------------------------------------
// main
// ---------------------------------------------------------------------------

async function main(argv = process.argv.slice(2)) {
    const { values: flags, errors: parseErrors } = parseArgs(argv, FLAG_SCHEMA);

    if (flags.help) {
        process.stdout.write(HELP_TEXT);
        return EXIT.OK;
    }

    const json = !!flags.json;
    const dryRun = !!flags.dryRun;
    // --json suppresses clack UI and therefore also disables prompts (otherwise
    // they'd corrupt the JSON stream). --non-interactive is the explicit knob.
    const nonInteractive = !!flags.nonInteractive || json;
    const brandRoot = process.cwd();
    const result = newResult({ dryRun, brandRoot });

    function fail(exitCode, errs) {
        result.status = 'error';
        result.errors.push(...errs);
        if (json) emitJson(result);
        else for (const e of errs) logger.error(e.message);
        return exitCode;
    }

    if (parseErrors.length > 0) {
        return fail(EXIT.INVALID_INPUT, parseErrors);
    }

    if (flags.slug) {
        const err = validateSlug(flags.slug);
        if (err) return fail(EXIT.INVALID_INPUT, [{ code: 'INVALID_INPUT', message: `--slug: ${err}` }]);
    }
    if (flags.keepAiContext && !flags.aiContext) {
        result.warnings.push('--keep-ai-context has no effect without --ai-context');
    }

    let ui;
    if (json) {
        ui = makeNoopUi();
    } else {
        const clack = await import('@clack/prompts');
        ui = makeClackUi(clack);
    }

    if (!json) logger.banner();
    if (!json && !nonInteractive) {
        ui.intro('Next Campaign Page Kit — start a campaign');
    } else if (!json) {
        logger.info('campaign-init (non-interactive)');
    }

    // 1. package.json scripts ------------------------------------------------
    const packageJsonPath = path.join(brandRoot, 'package.json');
    if (fs.existsSync(packageJsonPath)) {
        const current = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
        const { pkg, added } = mergeScripts(current);
        if (added.length > 0 && !dryRun) {
            fs.writeFileSync(packageJsonPath, JSON.stringify(pkg, null, 2) + '\n', 'utf8');
        }
        result.writes.package_json_scripts_added = added;
        if (!json) {
            if (added.length > 0) ui.log.success(`added package.json scripts: ${D}${added.join(', ')}${R}`);
            else ui.log.step('package.json scripts already present');
        }
    } else {
        result.warnings.push('no package.json found — run `npm init -y` first');
        if (!json) ui.log.warn(`no package.json found — run ${D}npm init -y${R} first`);
    }

    // 2. _data/campaigns.json -----------------------------------------------
    const campaignsDataPath = path.join(brandRoot, '_data', 'campaigns.json');
    if (!fs.existsSync(campaignsDataPath)) {
        if (!dryRun) {
            fs.mkdirSync(path.dirname(campaignsDataPath), { recursive: true });
            fs.writeFileSync(campaignsDataPath, CAMPAIGNS_JSON, 'utf8');
        }
        result.writes.campaigns_json_seeded = true;
        if (!json) ui.log.success('created _data/campaigns.json');
    } else if (!json) {
        ui.log.step('_data/campaigns.json already exists');
    }

    // 3. fetch templates.json -----------------------------------------------
    let manifest;
    try {
        manifest = await fetchJson(TEMPLATES_URL);
    } catch (err) {
        return fail(EXIT.UPSTREAM_FETCH_FAILED, [{
            code: 'UPSTREAM_FETCH_FAILED',
            message: `could not fetch templates.json: ${err.message}`,
        }]);
    }
    const templates = selectableTemplates(manifest);
    if (templates.length === 0) {
        return fail(EXIT.UPSTREAM_FETCH_FAILED, [{
            code: 'UPSTREAM_FETCH_FAILED',
            message: 'upstream templates.json had no selectable templates',
        }]);
    }

    // 4. resolve template slug ----------------------------------------------
    let templateSlug = flags.template;
    if (templateSlug) {
        if (!templates.find(t => t.slug === templateSlug)) {
            return fail(EXIT.UPSTREAM_TEMPLATE_NOT_FOUND, [{
                code: 'UPSTREAM_TEMPLATE_NOT_FOUND',
                message: `template "${templateSlug}" not found in upstream templates.json`,
                hint: `valid: ${templates.map(t => t.slug).join(', ')}`,
            }]);
        }
    } else if (nonInteractive) {
        return fail(EXIT.MISSING_INPUT, [{
            code: 'MISSING_INPUT',
            message: '--template is required in --non-interactive mode',
            hint: `valid: ${templates.map(t => t.slug).join(', ')}`,
        }]);
    } else {
        const picked = await ui.select({
            message: 'Pick a starter template',
            options: templates.map(t => ({
                value: t.slug,
                label: t.deprecated ? `[DEPRECATED] ${t.name}` : t.name,
                hint: t.description || t.slug,
            })),
        });
        if (ui.isCancel(picked)) { ui.outro('Cancelled.'); return EXIT.OK; }
        templateSlug = picked;
    }
    const pickedTemplate = templates.find(t => t.slug === templateSlug);

    // 5. resolve name -------------------------------------------------------
    let finalName = flags.name && String(flags.name).trim();
    if (!finalName) {
        if (nonInteractive) {
            finalName = pickedTemplate.name;
        } else {
            const entered = await ui.text({
                message: `Campaign name ${D}(display name shown in your dev server picker)${R}`,
                placeholder: pickedTemplate.name,
                initialValue: pickedTemplate.name,
                validate: (v) => { if (!(v || '').trim()) return 'name cannot be empty'; },
            });
            if (ui.isCancel(entered)) { ui.outro('Cancelled.'); return EXIT.OK; }
            finalName = entered.trim();
        }
    }

    // 6. resolve slug -------------------------------------------------------
    let finalSlug = flags.slug && String(flags.slug).trim();
    if (!finalSlug) {
        const defaultSlug = slugify(finalName) || templateSlug;
        if (nonInteractive) {
            finalSlug = defaultSlug;
        } else {
            const entered = await ui.text({
                message: `Campaign slug ${D}(directory name under src/ and URL path: e.g. /my-campaign/)${R}`,
                placeholder: defaultSlug,
                initialValue: defaultSlug,
                validate: (v) => validateSlug(v),
            });
            if (ui.isCancel(entered)) { ui.outro('Cancelled.'); return EXIT.OK; }
            finalSlug = entered.trim();
        }
    }

    // 7. conflict check -----------------------------------------------------
    const destDir = path.join(brandRoot, 'src', finalSlug);
    let localCampaigns;
    try {
        localCampaigns = JSON.parse(fs.readFileSync(campaignsDataPath, 'utf8'));
    } catch (err) {
        return fail(EXIT.GENERIC, [{
            code: 'GENERIC',
            message: `could not read _data/campaigns.json: ${err.message}`,
        }]);
    }
    const hasDirectory = fs.existsSync(destDir);
    const hasRegistryEntry = Object.prototype.hasOwnProperty.call(localCampaigns, finalSlug);
    if (hasDirectory || hasRegistryEntry) {
        const parts = [hasDirectory && `src/${finalSlug}/`, hasRegistryEntry && `_data/campaigns.json[${finalSlug}]`].filter(Boolean);
        if (nonInteractive) {
            if (!flags.overwrite) {
                return fail(EXIT.CONFLICT, [{
                    code: 'CONFLICT',
                    message: `${parts.join(' and ')} already exists`,
                    hint: 're-run with --overwrite to replace',
                }]);
            }
        } else {
            const ok = await ui.confirm({ message: `${parts.join(' and ')} already exists. Overwrite?`, initialValue: false });
            if (ui.isCancel(ok) || !ok) { ui.outro('Cancelled.'); return EXIT.OK; }
        }
    }

    // 8. fetch upstream registry entry (best effort) ------------------------
    let upstreamEntry = null;
    try {
        const upstream = await fetchJson(CAMPAIGNS_URL);
        upstreamEntry = upstream[templateSlug] || null;
        if (!upstreamEntry) {
            const msg = `upstream registry has no entry for ${templateSlug}; using a minimal local entry`;
            result.warnings.push(msg);
            if (!json) ui.log.warn(msg);
        }
    } catch (err) {
        const msg = `could not fetch upstream registry entry: ${err.message}; using a minimal local entry`;
        result.warnings.push(msg);
        if (!json) ui.log.warn(msg);
    }

    const nextCampaigns = mergeCampaignEntry(localCampaigns, finalSlug, upstreamEntry, finalName);

    result.campaign = {
        slug: finalSlug,
        name: finalName,
        template: templateSlug,
        sdk_version: (upstreamEntry && upstreamEntry.sdk_version) || null,
    };
    result.writes.campaigns_json_entry = finalSlug;
    result.writes.destination = `src/${finalSlug}/`;

    // 9. dry-run early exit -------------------------------------------------
    if (dryRun) {
        if (flags.aiContext && flags.aiContext !== 'none') {
            try {
                const written = await writeAiContext({
                    brandRoot, tool: flags.aiContext, keep: !!flags.keepAiContext, dryRun: true,
                });
                if (written) result.writes.ai_context.push({ tool: flags.aiContext, ...written });
            } catch (err) {
                result.warnings.push(`AI context (dry-run): ${err.message}`);
            }
        }
        if (json) emitJson(result);
        else if (!nonInteractive) ui.outro('Dry run — no writes performed.');
        else logger.info('dry run — no writes performed');
        return EXIT.OK;
    }

    // 10. download + extract + atomic replace -------------------------------
    const sp = ui.spinner();
    sp.start(`Downloading ${templateSlug}`);
    let stagingRoot;
    try {
        const destParent = path.dirname(destDir);
        fs.mkdirSync(destParent, { recursive: true });
        stagingRoot = fs.mkdtempSync(path.join(destParent, '.cpk-install-'));
        const stagedDir = path.join(stagingRoot, finalSlug);
        const tarball = await fetchBuffer(TARBALL_URL);
        const written = extractTemplate(tarball, templateSlug, stagedDir);
        replaceDirectoryWithRollback(stagedDir, destDir, () => {
            fs.writeFileSync(campaignsDataPath, JSON.stringify(nextCampaigns, null, 2) + '\n', 'utf8');
        });
        sp.stop(`Wrote ${written} files to src/${finalSlug}/`);
        result.writes.files_extracted = written;
    } catch (err) {
        sp.stop('Template install failed');
        return fail(EXIT.UPSTREAM_FETCH_FAILED, [{
            code: 'UPSTREAM_FETCH_FAILED',
            message: `template install failed: ${err.message}`,
        }]);
    } finally {
        if (stagingRoot) {
            fs.rmSync(stagingRoot, { recursive: true, force: true });
        }
    }
    if (!json) ui.log.success(`Registered ${D}${finalSlug}${R} in _data/campaigns.json`);

    // 11. API key -----------------------------------------------------------
    let apiKey = flags.apiKey && String(flags.apiKey).trim();
    if (!apiKey && !nonInteractive) {
        const wantsKey = await ui.confirm({ message: 'Set the Campaign API key now?', initialValue: true });
        if (!ui.isCancel(wantsKey) && wantsKey) {
            const entered = await ui.text({
                message: `API key ${D}(from your store's Campaigns App — written to assets/config.js)${R}`,
                placeholder: 'your-api-key',
                validate: (v) => { if (!(v || '').trim()) return 'API key cannot be empty'; },
            });
            if (!ui.isCancel(entered)) apiKey = String(entered).trim();
        }
    }
    if (apiKey) {
        const configPath = path.join(destDir, 'assets', 'config.js');
        try {
            const updated = applyApiKey(fs.readFileSync(configPath, 'utf8'), apiKey);
            if (updated === null) {
                const msg = 'no apiKey field found in assets/config.js — set it manually';
                result.warnings.push(msg);
                if (!json) ui.log.warn(msg);
            } else {
                fs.writeFileSync(configPath, updated, 'utf8');
                result.writes.api_key_applied = true;
                if (!json) ui.log.success('API key written to assets/config.js');
            }
        } catch (err) {
            const msg = `could not write API key: ${err.message}`;
            result.warnings.push(msg);
            if (!json) ui.log.warn(msg);
        }
    }

    // 12. AI context --------------------------------------------------------
    // Interactive flow:
    //   1. Ask "Set up AI context?"
    //   2. If yes, pick a tool
    //   3. If that tool's destination file already exists, confirm update
    // Non-interactive: only --ai-context <tool> drives this (no prompts).
    const aiContextTools = [];
    if (flags.aiContext && flags.aiContext !== 'none') {
        aiContextTools.push(flags.aiContext);
    } else if (!nonInteractive) {
        const wants = await ui.confirm({
            message: `Set up AI context now? ${D}(installs CLAUDE.md / AGENTS.md / etc. for your editor or agent)${R}`,
            initialValue: false,
        });
        if (!ui.isCancel(wants) && wants) {
            const picked = await ui.select({
                message: 'Which tool?',
                options: [
                    { value: 'claude',  label: 'Claude Code',    hint: 'CLAUDE.md' },
                    { value: 'codex',   label: 'OpenAI Codex',   hint: 'AGENTS.md' },
                    { value: 'cursor',  label: 'Cursor',         hint: '.cursor/rules/campaign-page-kit.mdc' },
                    { value: 'copilot', label: 'GitHub Copilot', hint: '.github/copilot-instructions.md' },
                ],
            });
            if (!ui.isCancel(picked)) {
                const target = aiContextTarget(picked);
                const exists = fs.existsSync(path.join(brandRoot, target.path));
                if (exists) {
                    const update = await ui.confirm({
                        message: `${target.path} already exists. Update it?`,
                        initialValue: false,
                    });
                    if (!ui.isCancel(update) && update) aiContextTools.push(picked);
                } else {
                    aiContextTools.push(picked);
                }
            }
        }
    }

    if (aiContextTools.length > 0) {
        for (const tool of aiContextTools) {
            try {
                const written = await writeAiContext({
                    brandRoot, tool, keep: !!flags.keepAiContext, dryRun: false,
                });
                if (written) {
                    result.writes.ai_context.push({ tool, ...written });
                    if (!json) ui.log.success(`AI context (${tool}): ${written.action} ${written.path}`);
                }
            } catch (err) {
                // Non-fatal: scaffold itself succeeded.
                const msg = `AI context (${tool}): ${err.message}`;
                result.warnings.push(msg);
                if (!json) ui.log.warn(msg);
            }
        }
    }

    // 13. emit final result -------------------------------------------------
    if (json) {
        emitJson(result);
    } else if (!nonInteractive) {
        ui.outro(`Next step: ${D}npm run dev${R} — preview ${finalSlug}`);
    } else {
        logger.info(`done — preview with: npm run dev (pick ${finalSlug})`);
    }

    return EXIT.OK;
}

if (require.main === module) {
    main().then(
        (code) => process.exit(code || 0),
        (err) => {
            logger.error(err && err.message || String(err));
            process.exit(EXIT.GENERIC);
        }
    );
}

module.exports = {
    SCRIPTS,
    EXIT,
    AI_CONTEXT_TARGETS,
    AI_CONTEXT_SENTINEL,
    FLAG_SCHEMA,
    slugify,
    selectableTemplates,
    mergeCampaignEntry,
    mergeScripts,
    applyApiKey,
    extractTemplate,
    countFiles,
    replaceDirectoryWithRollback,
    parseArgs,
    kebabToCamel,
    validateSlug,
    aiContextTarget,
    buildAiContextContent,
    main,
};
